### PR TITLE
Add cattagservice

### PR DESCRIPTION
**Pull request by georgemakadze**
1. I created CatListService file.
2. Add function in CatListServing.
3. Define catTagApiURL.

### CODE REVIEW

**Review by ashalva — DISMISSED** [review]
Perfect ✅

**Review by ashalva — COMMENTED** [review]
Let it be here for some time, lets see how it goes.

File: GeorgianMemenator/Scenes/MainView/MainViewModel.swift [hunk 1]
```diff
@@ -24,6 +24,7 @@ class MainViewModel: ObservableObject {
     
     init() {
         generateRandomMeme()
+
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/MainView/MainViewModel.swift]: Unneeded change

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift [hunk 2]
```diff
@@ -8,11 +8,57 @@
...
-    @Published var tags: [CatTagDetail] = [
-        CatTagDetail(name: "Sphynx cat"),
-        CatTagDetail(name: "Munchkin cat"),
-        CatTagDetail(name: "Bengal cat")
-    ]
+    private let catTagsService: CatListServing
+    
+    @Published var tags: [CatTag] = []
+    
+    @Published private(set) var hasError: Bool = false
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift]: This pr does not concentrate on loading, errors or image loading. Please delete uneeded properties

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift [hunk 3]
```diff
@@ -8,11 +8,57 @@
...
+    
+    deinit {
+        cancellables.removeAll()
+    }
+    
+    func initialFetch() {
+        getCatTags()
+    }
+    
+    func finishImageLoading() {
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift]: There is no image in cat tags list. Its copy paste issue from `RedditViewModel`

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListView.swift [hunk 4]
```diff
@@ -11,17 +11,25 @@ import Foundation
...
+        })
+        .edgesIgnoringSafeArea(.bottom)
     }
+    
 }
 
 struct CatTagsListView_Previews: PreviewProvider {
     static var previews: some View {
-        CatTagsListView(viewModel: CatTagsListViewModel())
+        MainView(
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListView.swift:28]: You should not have changed it with `MainView`

File: GeorgianMemenator/Network/Services/CatListService.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import Foundation
+import Combine
+
+
+protocol CatListServing {
+    func getCatTagss() -> AnyPublisher<[CatTag], Error>
+}
+
+class CatListService: CatListServing {
+    private let catTagApiURL = "https://cataas.com/api/tags"
```
Comment (ashalva) [inline GeorgianMemenator/Network/Services/CatListService.swift]: The url is not correct, remove tags, it should be `https://cataas.com/api`

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift [hunk 6]
```diff
@@ -8,11 +8,57 @@
...
-    @Published var tags: [CatTagDetail] = [
-        CatTagDetail(name: "Sphynx cat"),
-        CatTagDetail(name: "Munchkin cat"),
-        CatTagDetail(name: "Bengal cat")
-    ]
+    private let catTagsService: CatListServing
+    
+    @Published var tags: [CatTag] = []
+    
+    //@Published private(set) var hasError: Bool = false
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift]: Comments must be removed

**Review by ashalva — COMMENTED** [review]
I can not see if real cats are fetched from server. Can you upload image somewhere in the pr?

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift [hunk 7]
```diff
@@ -8,11 +8,47 @@
...
-    @Published var tags: [CatTagDetail] = [
-        CatTagDetail(name: "Sphynx cat"),
-        CatTagDetail(name: "Munchkin cat"),
-        CatTagDetail(name: "Bengal cat")
-    ]
+    private let catTagsService: CatListServing
+    
+    @Published var tags: [CatTag] = []
+    
+    @Published private(set) var isLoading: Bool = false
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift]: isLoading and imageLoader must be removed as well, plus image was not uploaded properly

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift [hunk 8]
```diff
@@ -8,11 +8,40 @@
...
+        getCatTags()
+    }
+    
+    func getCatTags() {
+        
+        catTagsService
+            .getCatTagss()
+            .receive(on: DispatchQueue.main)
+            .sink(
+            receiveCompletion: { [weak self] resp in
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListViewModel.swift]: Once you remove all logic from `recieveCompletion` closure, `[weak self]` is not needed any more. Can be safely removed. @georgemakadze

**Comment by georgemakadze** [comment]
I corrected the mistakes if you could see it

File: GeorgianMemenator/Scenes/CatTagsList/CatTagsListView.swift [hunk 9]
```diff
@@ -11,17 +11,24 @@ import Foundation
 
 struct CatTagsListView: View {
     @ObservedObject var viewModel: CatTagsListViewModel
+    @State var image: UIImage = UIImage()
```
Comment (ashalva) [inline GeorgianMemenator/Scenes/CatTagsList/CatTagsListView.swift]: There is no image in cat tags list.

**Review by ashalva — COMMENTED** [review]
One issue left before merging 👍

**Review by ashalva — APPROVED** [review]
Nice one 👍

**Pull request merged**